File: updated_code/code/python/1114-print-in-order.py
import threading

class Foo:
    def __init__(self):
        self.lock = threading.Lock()
        self.condition = threading.Condition(self.lock)
        self.done = set()
        self.done.add(0)

    def first(self, printFirst):
        with self.lock:
            printFirst()
            self.done.clear()
            self.done.add(1)
            self.condition.notify_all()

    def second(self, printSecond):
        with self.lock:
            while 1 not in self.done:
                self.condition.wait()
            printSecond()
            self.done.clear()
            self.done.add(2)
            self.condition.notify_all()

    def third(self, printThird):
        with self.lock:
            while 2 not in self.done:
                self.condition.wait()
            printThird()
            self.done.clear()
            self.done.add(3)

File: updated_code/code/java/1114-print-in-order.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Foo {
    private boolean firstFinished, secondFinished;
    private Lock lock = new ReentrantLock();
    private Condition con1, con2;

    public Foo() {
        lock.lock();
        try {
            con1 = lock.newCondition();
            con2 = lock.newCondition();
        } finally {
            lock.unlock();
        }
    }

    public void first(Runnable printFirst) throws InterruptedException {
        printFirst.run();
        lock.lock();  // To acquire the lock before writing
        try {
            firstFinished = true;
            con1.signal();  // Signal that first is finished
        } finally {
            lock.unlock();
        }
    }

    public void second(Runnable printSecond) throws InterruptedException {
        lock.lock();  // Acquire lock before waiting
        try {
            while (!firstFinished) {
                con1.await();  // Wait for first to finish
            }
            printSecond.run();
            secondFinished = true;
            con2.signal();  // Signal that second is finished
        } finally {
            lock.unlock();
        }
    }

    public void third(Runnable printThird) throws InterruptedException {
        lock.lock();  // Acquire lock before waiting
        try {
            while (!secondFinished) {
                con2.await();  // Wait for second to finish
            }
            printThird.run();
        } finally {
            lock.unlock();
        }
    }
}